src/pipe.rs

```diff
@@ -0,0 +1,54 @@
+use std::sync::Arc;
+
+use crate::channel::ChannelTx;
+use tokio::sync::{mpsc, Mutex};
+
+pub(crate) type PipeTx = mpsc::Sender<PipeMessage>;
+pub(crate) type PipeRx = mpsc::Receiver<PipeMessage>;
+pub(crate) type SharedPipe = Arc<Mutex<Option<Pipe>>>;
+
+/// # Bidirectional communication pipe for `SubscribeLoop`
+///
+/// `PubNub` owns a reference to one end of the pipe for communiating with the `SubscribeLoop`.
+/// `Subscription` owns a clone of the sending side of the pipe to the `SubscribeLoop`, but not the
+/// receiving side. `SubscribeLoop` is capable of sending messages to `PubNub`, and receiving
+/// messages from both `PubNub` and any number of `Subscription` streams.
+///
+/// See [`PipeMessage`] for more details.
+#[derive(Debug)]
+pub(crate) struct Pipe {
+    pub(crate) tx: PipeTx, // Send-side for bidirectional communication
+    pub(crate) rx: PipeRx, // Recv-side for bidirectional communication
+}
+
+/// # The kinds of messages allowed to be delivered over a `Pipe`
+#[derive(Debug)]
+pub(crate) enum PipeMessage {
```

Comment:
Maybe we could use two kinds of Pipe -- client to SubscribeLoop and SubscribeLoop to client? That could address the "only sent from" constraints.

Reply:
Yeah that thought crossed my mind, as well. Particularly as I was writing the doc comments for the PipeMessage variants.